### PR TITLE
Bug fixes and minor improvements

### PR DESCRIPTION
**Pull request by stevebleazard**
The following bug fixes have been included:
- add let _ = to_yojson / of_yojson to generated code to avoid warnings when they aren't used
- Fix bug where doing [@@deriving of_yojson] causes an unused rec warning

The following minor enhancement has been added
- ty_of_yojson_exn created to raise an exception rather than return an error (issue [#57](https://github.com/ocaml-ppx/ppx_deriving_yojson/issues/57))

### CODE REVIEW

**Comment by stevebleazard** [comment]
Looks like the travis build failed because the OS package update failed

**Comment by whitequark** [comment]
I strongly prefer that such changes not be squashed into a single commit, this makes it pretty hard to review.

**Comment by stevebleazard** [comment]
I'd love to split these out but have limited time at present, can you take a look?

**Comment by stevebleazard** [comment]
I have split out the major changes into 2 commits
- unused warmings
- _exn function enhancements

File: src/ppx_deriving_yojson.cppo.ml [hunk 1]
```diff
@@ -557,11 +561,24 @@ let desu_str_of_type ~options ~path ({ ptype_loc = loc } as type_decl) =
...
+    let { ptype_params; _ } = type_decl in
+    let var_s_exn_args = List.mapi (fun i _ -> argn i |> evar) ptype_params in
+    let var_s_exn_args = var_s_exn_args @ [evar "x"] in
+    let var_s_exn_fun =
+      let rec loop = function
+      | [] -> [%expr match  [%e app (evar var_s) var_s_exn_args] with Ok x -> x | Error err -> raise (Failure err)]
+      | hd::tl -> lam (pvar hd) (loop tl)
+      in
+      loop ((List.mapi (fun i _ -> argn i) ptype_params) @ ["x"])
+    in
```
Comment (whitequark) [inline src/ppx_deriving_yojson.cppo.ml]: Can you show me the generated code for the _exn parts?

**Comment by whitequark** [comment]
Your first commit looks fine. Your second commit broke CI.

File: src/ppx_deriving_yojson.cppo.ml (hunk 1 above)
Comment (stevebleazard) [reply src/ppx_deriving_yojson.cppo.ml]: Normal:
```ocaml
    let of_yojson_exn x =
      match of_yojson x with | Ok x -> x | Error err -> raise (Failure err) 
    let _ = of_yojson_exn 
```
Polymorphic:
```ocaml
    let of_yojson_exn arg0 arg1 x =
      match of_yojson arg0 arg1 x with
      | Ok x -> x
      | Error err -> raise (Failure err)
    let _ = of_yojson_exn
```

**Comment by stevebleazard** [comment]
I've fixed the travis issues, the output for _exn is now:
Normal:
```ocaml
let of_yojson_exn x =
  match of_yojson x with
  | Result.Ok x -> x
  | Result.Error err -> raise (Failure err)
let _ = of_yojson_exn
```
Polymorphic:
```ocaml
let of_yojson_exn arg0 arg1 x =
  match of_yojson arg0 arg1 x with
  | Result.Ok x -> x
  | Result.Error err -> raise (Failure err)
let _ = of_yojson_exn
```

**Comment by whitequark** [comment]
Please rebase and then I'll merge.

**Comment by stevebleazard** [comment]
I've rebased and compared the upstream with my fork

**Comment by whitequark** [comment]
What I mean by "rebase" here is to use the "interactive rebase" feature to clean up the commit history. The fact that you re-triggered Travis does not need to be permanently recorded in the history of the project...

**Comment by stevebleazard** [comment]
Thanks, I have now cleaned up the history and psuhed. Sorry, for the confusion, this is the first time I've done this kind of rebase

**Comment by whitequark** [comment]
You should also get rid of the "reverted changes" commit and collapse everything related to the "_exn" functions into one.

**Comment by stevebleazard** [comment]
This has been completed

**Comment by whitequark** [comment]
Thank you for the PR!

**Pull request merged**